src/Microsoft.CodeQuality.Analyzers.Exp/UnitTests/Reliability/DisposeObjectsBeforeLosingScopeTests.cs

```diff
@@ -288,6 +288,185 @@ End Sub
             GetBasicResultAt(13, 13, "Sub Test.M1(a As A)", "New A()"));
         }
 
+        [Fact]
+        public void OutAndRefParametersWithDisposableAssignment_NoDisposeCall_NoDiagnostic()
+        {
+            VerifyCSharp(@"
+using System;
+
+class A : IDisposable
+{
+    public void Dispose()
+    {
+        throw new NotImplementedException();
+    }
+}
+
+class Test
+{
+    void M1(ref A a1, out A a2)
+    {
+        a1 = new A();
+        a2 = new A();
+    }
+}
+");
+
+            VerifyBasic(@"
+Imports System
+
+Class A
+    Implements IDisposable
+    Public Sub Dispose() Implements IDisposable.Dispose
+        Throw New NotImplementedException()
+    End Sub
+End Class
+
+Class Test
+    Sub M1(ByRef a As A)
+        a = New A()
+    End Sub
+End Class");
+        }
+
+        [Fact]
+        public void OutDisposableArgument_NoDisposeCall_Diagnostic()
+        {
+            VerifyCSharp(@"
+using System;
+
+class A : IDisposable
+{
+    public void Dispose()
+    {
+        throw new NotImplementedException();
+    }
+}
+
+class Test
+{
+    void M1(out A param)
+    {
+        param = new A();
+    }
+
+    void M2(out A param2)
+    {
+        M3(out param2);
+    }
+
+    void M3(out A param3)
+    {
+        param3 = new A();
+    }
+
+    void Method()
+    {
+        A a;
+        M1(out a);
+        A local = a;
+        M1(out a);
+
+        M1(out var a2);
+
+        A a3;
+        M2(out a3);
+    }
+}
+",
+            // Test0.cs(32,12): warning CA2000: In method 'void Test.Method()', call System.IDisposable.Dispose on object created by 'out a' before all references to it are out of scope.
+            GetCSharpResultAt(32, 12, "void Test.Method()", "out a"),
+            // Test0.cs(34,12): warning CA2000: In method 'void Test.Method()', call System.IDisposable.Dispose on object created by 'out a' before all references to it are out of scope.
+            GetCSharpResultAt(34, 12, "void Test.Method()", "out a"),
+            // Test0.cs(36,12): warning CA2000: In method 'void Test.Method()', call System.IDisposable.Dispose on object created by 'out var a2' before all references to it are out of scope.
+            GetCSharpResultAt(36, 12, "void Test.Method()", "out var a2"),
+            // Test0.cs(39,12): warning CA2000: In method 'void Test.Method()', call System.IDisposable.Dispose on object created by 'out a3' before all references to it are out of scope.
+            GetCSharpResultAt(39, 12, "void Test.Method()", "out a3"));
+        }
+
+        [Fact]
+        public void OutDisposableArgument_DisposeCall_NoDiagnostic()
+        {
+            VerifyCSharp(@"
+using System;
+
+class A : IDisposable
+{
+    public void Dispose()
+    {
+        throw new NotImplementedException();
+    }
+}
+
+class Test
+{
+    //void M1(out A param)
+    //{
+    //    param = new A();
+    //}
```

Comment:
I can assume this commented code is added intentionally to check that DFA does not cover it. If so, it would be good to support this with a comment.

Reply:
Thanks for the catch. Actually, I was testing the uncommented code in the test in isolation as it was failing. I forget to revert the temporary commenting out - will fix it now.